engine/collection/epochmgr/engine.go

```diff
@@ -207,32 +231,53 @@ func (e *Engine) onEpochTransition(first *flow.Header) error {
 
 	log.Info().Msg("epoch transition: new epoch components started successfully")
 
-	// Finally we set up a trigger to shut down the components of the previous
-	// epoch once we can no longer produce valid collections there. Transactions
-	// referencing blocks from the previous epoch are only valid for inclusion
-	// in collections built by clusters from that epoch. Consequently, it remains
-	// possible for the previous epoch's cluster to produce valid collections
-	// until all such transactions have expired. In fact, since these transactions
-	// can NOT be included by clusters in the new epoch, we MUST continue producing
-	// these collections within the previous epoch's clusters.
-	e.heightEvents.OnHeight(first.Height+flow.DefaultTransactionExpiry, func() {
+	// set up callback to stop previous epoch
+	e.prepareToStopEpochComponents(counter, lastEpochMaxHeight)
+
+	return nil
+}
+
+// prepareToStopEpochComponents registers a callback to stop the epoch with the
+// given counter once it is no longer possible to receive transactions from that
+// epoch. This occurs when we finalize sufficiently many blocks in the new epoch
+// that a transaction referencing any block from the previous epoch would be
+// considered immediately expired.
+//
+// Transactions referencing blocks from the previous epoch are only valid for
+// inclusion in collections built by clusters from that epoch. Consequently, it
+// remains possible for the previous epoch's cluster to produce valid collections
+// until all such transactions have expired. In fact, since these transactions
+// can NOT be included by clusters in the new epoch, we MUST continue producing
+// these collections within the previous epoch's clusters.
+//
+func (e *Engine) prepareToStopEpochComponents(epochCounter, epochMaxHeight uint64) {
+
+	stopAtHeight := epochMaxHeight + flow.DefaultTransactionExpiry + 1
```

Comment:
is this flow.DefaultTransactionExpiry enough, I remember we had some few extra buffer at collection time as well

Reply:
The buffer at collection build time is in the opposite direction. For a buffer value `b` if we have `t<=b` blocks of time to include a transaction, we will just throw it out because we know it's very unlikely for the collection to be included in a block before `b` blocks are built. If the collection was immediately included in a block, it would be valid, but we know there is usually a certain latency so we proactively discard "nearly-expired" transactions to make fully expired collections less likely.

Here, we are saying, once we observe a block with height such that a transaction referencing _any_ block in the previous epoch would be expired, then stop cluster consensus for this epoch.